[Modern Media Controls] remove window event listeners at deinitialization

calvaris opened the pull request:
#### a828da7eb601a0b75a4009529e3c439a6db7311f
<pre>
[Modern Media Controls] remove window event listeners at deinitialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=280503">https://bugs.webkit.org/show_bug.cgi?id=280503</a>

Reviewed by Ryan Reno.

DOMWindow keeps media controls object alive because of installed event listeners (see
JSDOMWindow::visitAdditionalChildren). This results in extra memory usage undil DOMWindow cleared (on top level
navigation).

Original patch by Eugene Mutavchi &lt;Ievgen_Mutavchi@comcast.com&gt;.

* Source/WebCore/Modules/modern-media-controls/controls/media-controls.js:
(MediaControls.prototype.reenable):
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.deinitialize):
(MediaController.prototype.reinitialize):

Canonical link: <a href="https://commits.webkit.org/284440@main">https://commits.webkit.org/284440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b1012376a0989cdc73a83571db64c2123d6a0101

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69140 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21812 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73221 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20298 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71257 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56341 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55027 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13473 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72206 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44304 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59688 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35506 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40972 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 4 new passes 4 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17118 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18672 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62915 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17463 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74932 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13122 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62680 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62581 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15395 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10579 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4190 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44344 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45417 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46613 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45159 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eocanha commented on Source/WebCore/Modules/modern-media-controls/controls/media-controls.js:
```diff
@@ -187,6 +187,11 @@ class MediaControls extends LayoutNode
 
+    reenable()
+    {
+        window.addEventListener("dragstart", this, true);
```
This reenable() method isn't symetric to the disable() method. It's missing the reinstall of the "focusin" event listener, which is installed at creation time and uninstalled by disable() [here](https://github.com/WebKit/WebKit/pull/34350/commits/67bab4d4eee5640301ffc3c3a5132c7b50d3fbcc#diff-c73d8fa8d6e6c156a9d1b1871e8db6ca63a2686912354757011556091aff568fL75).

calvaris replied:
You're right, I will submit something soon. I also moved the reenable function to a place closer to the disable one.

eocanha reviewed (commented):
I'm not a WebKit reviewer, but as author of https://commits.webkit.org/277196@main, I have some idea about this kind of problems, so I've left a comment.
If I was a reviewer, the comment was addressed and there were no regressions in EWS, I would r+ this PR.
A last thing: Please add "See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1400" in the commit message and in https://bugs.webkit.org/show_bug.cgi?id=280503. Also add "Original author: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>" in the commit message. It helps traceability.

rreno reviewed (approved):
r=me given @eocanha 's informal review.

calvaris committed:
[Modern Media Controls] remove window event listeners at deinitialization
https://bugs.webkit.org/show_bug.cgi?id=280503

Reviewed by Ryan Reno.

DOMWindow keeps media controls object alive because of installed event listeners (see
JSDOMWindow::visitAdditionalChildren). This results in extra memory usage undil DOMWindow cleared (on top level
navigation).

Original patch by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>.

* Source/WebCore/Modules/modern-media-controls/controls/media-controls.js:
(MediaControls.prototype.reenable):
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.deinitialize):
(MediaController.prototype.reinitialize):

Canonical link: https://commits.webkit.org/284440@main
